Sources/LambdaConverters.Wpf/Validator.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Diagnostics;
+using System.Globalization;
+using System.Windows.Controls;
+using JetBrains.Annotations;
+
+namespace LambdaConverters
+{
+    /// <summary>
+    /// A factory class used to create lambda-based instances of the <see cref="ValidationRule"/> class.
+    /// </summary>
+    public static class Validator
+    {
+        abstract class Rule : LambdaConverters.Rule
+        {
+            protected Rule(
+                RuleErrorStrategy errorStrategy,
+                object defaultInputTypeValue,
+                [NotNull] Type inputType,
+                bool isRuleFunctionAvailable)
+                : base(
+                    errorStrategy,
+                    defaultInputTypeValue,
+                    inputType,
+                    isRuleFunctionAvailable) { }
+
+            protected abstract ValidationResult ValidateInternal(object value, CultureInfo cultureInfo);
+
+            public override ValidationResult Validate(object value, CultureInfo cultureInfo)
+            {
+                if (!IsRuleFunctionAvailable)
+                {
+                    EventSource.Log.MissingRuleFunction("ruleFunction", ErrorStrategy.ToString());
+
+                    return GetErrorValue(default(ValidationResult));
+                }
+
+                return ValidateInternal(value, cultureInfo);
+            }
+        }
+
+        sealed class Rule<I> : Rule
+        {
+            readonly Func<ValidationRuleArgs<I>, ValidationResult> ruleFunction;
+
+            internal Rule(
+                Func<ValidationRuleArgs<I>, ValidationResult> ruleFunction,
+                RuleErrorStrategy errorStrategy)
+                : base(errorStrategy, default(I), typeof(I), ruleFunction != null)
+            {
+                this.ruleFunction = ruleFunction;
+            }
+
+            protected override ValidationResult ValidateInternal(object item, CultureInfo cultureInfo)
+            {
+                I inputValue;
+                try
+                {
+                    inputValue = (I)item;
+                }
+                catch (SystemException e) when (e is InvalidCastException || e is NullReferenceException)
+                {
+                    EventSource.Log.UnableToCastToRuleInputType(item?.GetType().Name ?? "null", typeof(I).Name, ErrorStrategy.ToString());
+
+                    return GetErrorValue(default(ValidationResult));
+                }
+
+                Debug.Assert(ruleFunction != null);
+
+                return ruleFunction(new ValidationRuleArgs<I>(inputValue, cultureInfo));
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ValidationRule" /> class.
+        /// </summary>
+        /// <typeparam name="I">The value type.</typeparam>
+        /// <param name="ruleFunction">The <see cref="ValidationRule.Validate" /> method.</param>
```

Comment:
Compiler warning CS0419: Ambiguous reference in cref attribute: 'ValidationRule.Validate'. Assuming 'ValidationRule.Validate(object, CultureInfo)', but could have also matched other overloads including 'ValidationRule.Validate(object, CultureInfo, BindingExpressionBase)'.

Please resolve.